### PR TITLE
fix(security): #EVAL-470 improve security on APIs of utilsController

### PR DESCRIPTION
**Pull request by Gregprod33**
## Describe your changes
- On route @Get("/types") : add ResourceFilter to check the structure
- On route @Get("/mainteachers/:idStructure") : add ResourceFilter to check the structure
- On route @Get("/enfants") : add resource filter AccessStructureIsParent to verify the structure and the the type of user (must be "Relative")
- On route @Post("/link/check/data/classes") : add resource filter AccessParamLinkGroupCycleStructure to check structure and the user's right of type "competences.paramLinkGroupCycle". The front file linkGroupCycle.ts has been modified to pass structure id on query param.
- On route @Post("/eleve/evenements"): set workflow right "can_update_retards_and_absences"
- On route @Post("/graph/img"): set workflow right "can_access_export_bulletin"

Finally, i created unit tests for the AccessParamLinkGroupCycleStructure.class

## Checklist tests
For the route @Post("/link/check/data/classes"), go to viescolaire, competences thumbnail and "rattachement des cycles" to bind a new cycle to a class.
## Issue ticket number and link
[#EVAL-470](https://jira.support-ent.fr/browse/EVAL-470)
## Checklist before requesting a review (magic string, indentation, comment/documentation...)

- [x] I have detailed the tests to do in my feature/fix in order to prevent consequents regressions (must specify in **Checklist tests**)
- [x] I have performed a self-review of my code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation (API Doc etc...) - (must specify in **Description** for target version)
- [ ] If it is a consequent feature, I have added thorough tests.
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] Any dependent changes have been added to this project (must specify in **Description**)

### CODE REVIEW

File: src/main/java/fr/openent/competences/controllers/UtilsController.java [hunk 1]
```diff
@@ -91,7 +92,8 @@ public void handle(UserInfos user) {
      */
     @Get("/mainteachers/:idStructure")
     @ApiDoc("Retourne tous les types de devoir par etablissement")
-    @SecuredAction(value = "", type = ActionType.AUTHENTICATED)
+    @SecuredAction(value = "", type = ActionType.RESOURCE)
+    @ResourceFilter(AccessIfMyStructure.class)
     public void viewTittulaires(final HttpServerRequest request) {
```
Comment (lapeyronnie) [inline src/main/java/fr/openent/competences/controllers/UtilsController.java:97]: Peux-tu changer la doc de cette API, stp par "Récupère la liste des professeurs titulaires d'un remplaçant sur un établissement donné".
Merci

File: src/main/java/fr/openent/competences/controllers/UtilsController.java [hunk 2]
```diff
@@ -279,7 +284,7 @@ public void handle(final JsonObject ressource) {
     }
 
     @Post("/graph/img")
-    @SecuredAction(value ="", type = ActionType.AUTHENTICATED)
+    @SecuredAction(Competences.CAN_ACCESS_EXPORT_BULLETIN)
     public void postImgForBulletins(final HttpServerRequest request){
```
Comment (lapeyronnie) [inline src/main/java/fr/openent/competences/controllers/UtilsController.java:288]: Il ne faut ppas créer le droit mais le vérifer => ressourceFilter

**Review by lapeyronnie — CHANGES_REQUESTED** [review]
add filterRessource

File: src/main/java/fr/openent/competences/controllers/UtilsController.java (hunk 1 above)
Comment (Gregprod33) [reply src/main/java/fr/openent/competences/controllers/UtilsController.java:97]: ok

File: src/main/java/fr/openent/competences/controllers/UtilsController.java (hunk 2 above)
Comment (Gregprod33) [reply src/main/java/fr/openent/competences/controllers/UtilsController.java:290]: rectifié aussi

File: src/main/java/fr/openent/competences/security/AccessParamLinkGroupCycleStructure.java [hunk 3]
```diff
@@ -0,0 +1,21 @@
...
+
+public class AccessParamLinkGroupCycleStructure implements ResourcesProvider {
+    @Override
+    public void authorize(HttpServerRequest request, Binding binding, UserInfos user, Handler<Boolean> handler) {
+        String structureId = WorkflowActionUtils.getParamStructure(request);
+        if(structureId == null) {
+            handler.handle(false);
+        } else {
+            handler.handle(user.getStructures().contains(structureId) && WorkflowActionUtils.hasRight(user, Competences.PARAM_LINK_GROUP_CYCLE_RIGHT.toString()));
+        }
```
Comment (qpic-cgi) [inline src/main/java/fr/openent/competences/security/AccessParamLinkGroupCycleStructure.java:19]: handler.handle(
structureId != null &&
user.getStructures().contains(structureId) && 
WorkflowActionUtils.hasRight(user, Competences.PARAM_LINK_GROUP_CYCLE_RIGHT.toString()));

ça évite de faire des gros if/else

File: src/main/java/fr/openent/competences/security/AccessStructureIsParent.java [hunk 4]
```diff
@@ -0,0 +1,25 @@
...
+
+public class AccessStructureIsParent implements ResourcesProvider {
+    @Override
+    public void authorize(HttpServerRequest request, Binding binding, UserInfos user, Handler<Boolean> handler) {
+        String structureId = WorkflowActionUtils.getParamStructure(request);
+        if (structureId == null | !user.getType().equals(Field.RELATIVE)) {
+            handler.handle(false);
+        } else {
+            handler.handle(user.getStructures().contains(structureId));
+        }
```
Comment (qpic-cgi) [inline src/main/java/fr/openent/competences/security/AccessStructureIsParent.java:23]: CF commentaire au dessus

**Review by qpic-cgi — APPROVED** [review]

File: src/main/java/fr/openent/competences/security/AccessParamLinkGroupCycleStructure.java (hunk 3 above)
Comment (Gregprod33) [reply src/main/java/fr/openent/competences/security/AccessParamLinkGroupCycleStructure.java:19]: oui tout à fait, merci !

**Review by lapeyronnie — APPROVED** [review]

**Pull request merged**